From e81eb040ee71b08ecd27fdfd260b593707f517d7 Mon Sep 17 00:00:00 2001
From: jasshsu <jasshsu@ami.com>
Date: Tue, 14 Jul 2026 19:54:00 +0530
Subject: [PATCH] Add the M Hold patch and I2C Driver change from INTEL

Patch recreated by Vijay K S <vijayks@ami.com> for kernel upgrade v6.18

Upstream-Status: Inappropriate [AMI: driver changes]

---
 drivers/i2c/Kconfig                        |   6 +
 drivers/i2c/i2c-core-base.c                |  75 ++++++-
 drivers/i2c/i2c-core-smbus.c               |  20 +-
 drivers/i2c/i2c-mux.c                      | 151 +++++++++++---
 drivers/i2c/muxes/i2c-arb-gpio-challenge.c | 132 ++++++++++++-
 include/linux/device/class.h               |  23 +++
 include/linux/i2c-aspeed.h                 | 216 +++++++++++++++++++++
 include/linux/i2c-mux.h                    |   5 +
 include/linux/i2c.h                        |  28 ++-
 include/uapi/linux/i2c.h                   |   1 +
 10 files changed, 625 insertions(+), 32 deletions(-)
 create mode 100644 include/linux/i2c-aspeed.h

diff --git a/drivers/i2c/Kconfig b/drivers/i2c/Kconfig
index 32c34a426413..1db8cc32e1e3 100644
--- a/drivers/i2c/Kconfig
+++ b/drivers/i2c/Kconfig
@@ -185,6 +185,12 @@ config I2C_DEBUG_BUS
 	  a problem with I2C support and want to see more of what is going
 	  on.
 
+config DRAGON_CHASSIS
+	bool "Nvidia Dragon Chassis support"
+	help
+	  This enables all the necessary code for the support of the
+	  Dragon Chassis features.
+
 endif # I2C
 
 endmenu
diff --git a/drivers/i2c/i2c-core-base.c b/drivers/i2c/i2c-core-base.c
index 3039af033ad9..2f010a187c07 100644
--- a/drivers/i2c/i2c-core-base.c
+++ b/drivers/i2c/i2c-core-base.c
@@ -102,6 +102,26 @@ const char *i2c_freq_mode_string(u32 bus_freq_hz)
 }
 EXPORT_SYMBOL_GPL(i2c_freq_mode_string);
 
+static void i2c_adapter_hold(struct i2c_adapter *adapter, unsigned long timeout)
+{
+	mutex_lock(&adapter->hold_lock);
+	schedule_delayed_work(&adapter->unhold_work, timeout);
+}
+
+static void i2c_adapter_unhold(struct i2c_adapter *adapter)
+{
+	cancel_delayed_work_sync(&adapter->unhold_work);
+	mutex_unlock(&adapter->hold_lock);
+}
+
+static void i2c_adapter_unhold_work(struct work_struct *work)
+{
+	struct delayed_work *dwork = to_delayed_work(work);
+	struct i2c_adapter *adapter = container_of(dwork, struct i2c_adapter, unhold_work);
+
+	mutex_unlock(&adapter->hold_lock);
+}
+
 const struct i2c_device_id *i2c_match_id(const struct i2c_device_id *id,
 						const struct i2c_client *client)
 {
@@ -1627,6 +1647,7 @@ static int i2c_register_adapter(struct i2c_adapter *adap)
 
 	dev_dbg(&adap->dev, "adapter [%s] registered\n", adap->name);
 
+	mutex_init(&adap->hold_lock);
 	/* create pre-declared device nodes */
 	of_i2c_register_devices(adap);
 	i2c_acpi_install_space_handler(adap);
@@ -1639,6 +1660,7 @@ static int i2c_register_adapter(struct i2c_adapter *adap)
 	mutex_lock(&core_lock);
 	bus_for_each_drv(&i2c_bus_type, NULL, adap, __process_new_adapter);
 	mutex_unlock(&core_lock);
+	INIT_DELAYED_WORK(&adap->unhold_work, i2c_adapter_unhold_work);
 
 	return 0;
 
@@ -1856,6 +1878,8 @@ void i2c_del_adapter(struct i2c_adapter *adap)
 	idr_remove(&i2c_adapter_idr, adap->nr);
 	mutex_unlock(&core_lock);
 
+	i2c_adapter_unhold(adap);
+
 	/* Clear the device structure in case this adapter is ever going to be
 	   added again */
 	memset(&adap->dev, 0, sizeof(adap->dev));
@@ -2258,7 +2282,9 @@ static int i2c_check_for_quirks(struct i2c_adapter *adap, struct i2c_msg *msgs,
  */
 int __i2c_transfer(struct i2c_adapter *adap, struct i2c_msg *msgs, int num)
 {
+	enum i2c_hold_msg_type hold_msg = I2C_HOLD_MSG_NONE;
 	unsigned long orig_jiffies;
+	unsigned long timeout;
 	int ret, try;
 
 	if (!adap->algo->master_xfer) {
@@ -2276,6 +2302,25 @@ int __i2c_transfer(struct i2c_adapter *adap, struct i2c_msg *msgs, int num)
 	if (adap->quirks && i2c_check_for_quirks(adap, msgs, num))
 		return -EOPNOTSUPP;
 
+			/* Do not deliver a mux hold msg to root bus adapter */
+	if (!i2c_parent_is_i2c_adapter(adap)) {
+	    hold_msg = i2c_check_hold_msg(msgs[num - 1].flags,
+					   msgs[num - 1].len,
+					  (u16 *)msgs[num - 1].buf);
+		if (hold_msg == I2C_HOLD_MSG_SET) {
+			timeout = msecs_to_jiffies(*(u16 *)msgs[num - 1].buf);
+			i2c_adapter_hold(adap, timeout);
+
+			if (--num == 0)
+				return 0;
+		} else if (hold_msg == I2C_HOLD_MSG_RESET) {
+			i2c_adapter_unhold(adap);
+			return 0;
+		} else if (hold_msg == I2C_HOLD_MSG_NONE) {
+			mutex_lock(&adap->hold_lock);
+		}
+	}
+
 	/*
 	 * i2c_trace_msg_key gets enabled when tracepoint i2c_transfer gets
 	 * enabled.  This is an efficient way of keeping the for-loop from
@@ -2312,6 +2357,13 @@ int __i2c_transfer(struct i2c_adapter *adap, struct i2c_msg *msgs, int num)
 		trace_i2c_result(adap, num, ret);
 	}
 
+	if (!i2c_parent_is_i2c_adapter(adap)) {
+		if (hold_msg == I2C_HOLD_MSG_SET && ret < 0)
+			i2c_adapter_unhold(adap);
+		else if (hold_msg == I2C_HOLD_MSG_NONE)
+			mutex_unlock(&adap->hold_lock);
+	}
+
 	return ret;
 }
 EXPORT_SYMBOL(__i2c_transfer);
@@ -2330,6 +2382,7 @@ EXPORT_SYMBOL(__i2c_transfer);
  */
 int i2c_transfer(struct i2c_adapter *adap, struct i2c_msg *msgs, int num)
 {
+	bool do_bus_lock = true;
 	int ret;
 
 	/* REVISIT the fault reporting model here is weak:
@@ -2348,12 +2401,26 @@ int i2c_transfer(struct i2c_adapter *adap, struct i2c_msg *msgs, int num)
 	 *    one (discarding status on the second message) or errno
 	 *    (discarding status on the first one).
 	 */
-	ret = __i2c_lock_bus_helper(adap);
-	if (ret)
-		return ret;
+
+	/*
+	 * Do not lock a bus for delivering an unhold msg to a mux
+	 * adpater. This is just for a single length unhold msg case.
+	 */
+	if (num == 1 && i2c_parent_is_i2c_adapter(adap) &&
+	    i2c_check_hold_msg(msgs[0].flags, msgs[0].len,
+			       (u16 *)msgs[0].buf) ==
+			       I2C_HOLD_MSG_RESET)
+		do_bus_lock = false;
+
+	if (do_bus_lock) {
+		ret = __i2c_lock_bus_helper(adap);
+		if (ret)
+			return ret;
+	}
 
 	ret = __i2c_transfer(adap, msgs, num);
-	i2c_unlock_bus(adap, I2C_LOCK_SEGMENT);
+	if (do_bus_lock)
+		i2c_unlock_bus(adap, I2C_LOCK_SEGMENT);
 
 	return ret;
 }
diff --git a/drivers/i2c/i2c-core-smbus.c b/drivers/i2c/i2c-core-smbus.c
index 71eb1ef56f0c..1b60dc2b92e5 100644
--- a/drivers/i2c/i2c-core-smbus.c
+++ b/drivers/i2c/i2c-core-smbus.c
@@ -537,15 +537,33 @@ s32 i2c_smbus_xfer(struct i2c_adapter *adapter, u16 addr,
 		   unsigned short flags, char read_write,
 		   u8 command, int protocol, union i2c_smbus_data *data)
 {
+	bool do_bus_lock = true;
 	s32 res;
 
 	res = __i2c_lock_bus_helper(adapter);
 	if (res)
 		return res;
 
+	/*
+	 * Do not lock a bus for delivering an unhold msg to a mux adpater.
+	 * This is just for a single length unhold msg case.
+	 */
+	if (i2c_parent_is_i2c_adapter(adapter) &&
+	    i2c_check_hold_msg(flags,
+			       protocol == I2C_SMBUS_WORD_DATA ? 2 : 0,
+			       &data->word) == I2C_HOLD_MSG_RESET)
+		do_bus_lock = false;
+
+	if (do_bus_lock) {
+		res = __i2c_lock_bus_helper(adapter);
+		if (res)
+			return res;
+	}
+
 	res = __i2c_smbus_xfer(adapter, addr, flags, read_write,
 			       command, protocol, data);
-	i2c_unlock_bus(adapter, I2C_LOCK_SEGMENT);
+	if (do_bus_lock)
+		i2c_unlock_bus(adapter, I2C_LOCK_SEGMENT);
 
 	return res;
 }
diff --git a/drivers/i2c/i2c-mux.c b/drivers/i2c/i2c-mux.c
index 2c7ddfae55b5..e156cfa54a8e 100644
--- a/drivers/i2c/i2c-mux.c
+++ b/drivers/i2c/i2c-mux.c
@@ -27,6 +27,8 @@
 #include <linux/of.h>
 #include <linux/slab.h>
 #include <linux/sysfs.h>
+#include <linux/mutex.h>
+#include <linux/workqueue.h>
 
 /* multiplexer per channel data */
 struct i2c_mux_priv {
@@ -36,84 +38,180 @@ struct i2c_mux_priv {
 	u32 chan_id;
 };
 
+static void i2c_mux_hold(struct i2c_mux_core *muxc, u32 chan_id, unsigned long timeout)
+{
+	mutex_lock(&muxc->hold_lock);
+	muxc->holder_chan_id = chan_id;
+	schedule_delayed_work(&muxc->unhold_work, timeout);
+}
+
+static void i2c_mux_unhold(struct i2c_mux_core *muxc)
+{
+	cancel_delayed_work_sync(&muxc->unhold_work);
+	mutex_unlock(&muxc->hold_lock);
+}
+
+static void i2c_mux_unhold_work(struct work_struct *work)
+{
+	struct delayed_work *dwork = to_delayed_work(work);
+	struct i2c_mux_core *muxc = container_of(dwork, struct i2c_mux_core, unhold_work);
+
+	if (muxc->deselect)
+		muxc->deselect(muxc, muxc->holder_chan_id);
+	mutex_unlock(&muxc->hold_lock);
+}
+
 static int __i2c_mux_master_xfer(struct i2c_adapter *adap,
 				 struct i2c_msg msgs[], int num)
 {
 	struct i2c_mux_priv *priv = adap->algo_data;
 	struct i2c_mux_core *muxc = priv->muxc;
 	struct i2c_adapter *parent = muxc->parent;
+	enum i2c_hold_msg_type hold_msg;
+	unsigned long timeout;
 	int ret;
 
 	/* Switch to the right mux port and perform the transfer. */
 
+	hold_msg = i2c_check_hold_msg(msgs[num - 1].flags,
+				      msgs[num - 1].len,
+				      (u16 *)msgs[num - 1].buf);
+	if (hold_msg == I2C_HOLD_MSG_SET) {
+		timeout = msecs_to_jiffies(*(u16 *)msgs[num - 1].buf);
+		i2c_mux_hold(muxc, priv->chan_id, timeout);
+	} else if (hold_msg == I2C_HOLD_MSG_NONE) {
+		mutex_lock(&muxc->hold_lock);
+	}
+
 	ret = muxc->select(muxc, priv->chan_id);
 	if (ret >= 0)
 		ret = __i2c_transfer(parent, msgs, num);
-	if (muxc->deselect)
-		muxc->deselect(muxc, priv->chan_id);
+	if (hold_msg != I2C_HOLD_MSG_SET) {
+		if (muxc->deselect)
+			muxc->deselect(muxc, priv->chan_id);
+		if (hold_msg == I2C_HOLD_MSG_RESET)
+			i2c_mux_unhold(muxc);
+		else
+			mutex_unlock(&muxc->hold_lock);
+	} else if (hold_msg == I2C_HOLD_MSG_SET && ret < 0) {
+		i2c_mux_unhold(muxc);
+	}
 
 	return ret;
 }
 
 static int i2c_mux_master_xfer(struct i2c_adapter *adap,
-			       struct i2c_msg msgs[], int num)
+			      struct i2c_msg msgs[], int num)
 {
 	struct i2c_mux_priv *priv = adap->algo_data;
 	struct i2c_mux_core *muxc = priv->muxc;
 	struct i2c_adapter *parent = muxc->parent;
+	enum i2c_hold_msg_type hold_msg;
+	unsigned long timeout;
 	int ret;
 
 	/* Switch to the right mux port and perform the transfer. */
 
+	hold_msg = i2c_check_hold_msg(msgs[num - 1].flags,
+				      msgs[num - 1].len,
+				      (u16 *)msgs[num - 1].buf);
+	if (hold_msg == I2C_HOLD_MSG_SET) {
+		timeout = msecs_to_jiffies(*(u16 *)msgs[num - 1].buf);
+		i2c_mux_hold(muxc, priv->chan_id, timeout);
+	} else if (hold_msg == I2C_HOLD_MSG_NONE) {
+		mutex_lock(&muxc->hold_lock);
+	}
+
 	ret = muxc->select(muxc, priv->chan_id);
 	if (ret >= 0)
 		ret = i2c_transfer(parent, msgs, num);
-	if (muxc->deselect)
-		muxc->deselect(muxc, priv->chan_id);
+	if (hold_msg != I2C_HOLD_MSG_SET) {
+		if (muxc->deselect)
+			muxc->deselect(muxc, priv->chan_id);
+		if (hold_msg == I2C_HOLD_MSG_RESET)
+			i2c_mux_unhold(muxc);
+		else
+			mutex_unlock(&muxc->hold_lock);
+	}
 
 	return ret;
 }
 
 static int __i2c_mux_smbus_xfer(struct i2c_adapter *adap,
-				u16 addr, unsigned short flags,
-				char read_write, u8 command,
-				int size, union i2c_smbus_data *data)
+				 u16 addr, unsigned short flags,
+				 char read_write, u8 command,
+				 int size, union i2c_smbus_data *data)
 {
 	struct i2c_mux_priv *priv = adap->algo_data;
 	struct i2c_mux_core *muxc = priv->muxc;
 	struct i2c_adapter *parent = muxc->parent;
+	enum i2c_hold_msg_type hold_msg;
+	unsigned long timeout;
 	int ret;
 
 	/* Select the right mux port and perform the transfer. */
 
+	hold_msg = i2c_check_hold_msg(flags,
+				      size == I2C_SMBUS_WORD_DATA ? 2 : 0,
+				      &data->word);
+	if (hold_msg == I2C_HOLD_MSG_SET) {
+		timeout = msecs_to_jiffies(data->word);
+		i2c_mux_hold(muxc, priv->chan_id, timeout);
+	} else if (hold_msg == I2C_HOLD_MSG_NONE) {
+		mutex_lock(&muxc->hold_lock);
+	}
 	ret = muxc->select(muxc, priv->chan_id);
 	if (ret >= 0)
 		ret = __i2c_smbus_xfer(parent, addr, flags,
-				       read_write, command, size, data);
-	if (muxc->deselect)
-		muxc->deselect(muxc, priv->chan_id);
+				   read_write, command, size, data);
+	if (hold_msg != I2C_HOLD_MSG_SET) {
+		if (muxc->deselect)
+			muxc->deselect(muxc, priv->chan_id);
+		if (hold_msg == I2C_HOLD_MSG_RESET)
+			i2c_mux_unhold(muxc);
+		else
+			mutex_unlock(&muxc->hold_lock);
+	}
 
 	return ret;
 }
 
 static int i2c_mux_smbus_xfer(struct i2c_adapter *adap,
-			      u16 addr, unsigned short flags,
-			      char read_write, u8 command,
-			      int size, union i2c_smbus_data *data)
+			    u16 addr, unsigned short flags,
+			    char read_write, u8 command,
+			    int size, union i2c_smbus_data *data)
 {
 	struct i2c_mux_priv *priv = adap->algo_data;
 	struct i2c_mux_core *muxc = priv->muxc;
 	struct i2c_adapter *parent = muxc->parent;
+	enum i2c_hold_msg_type hold_msg;
+	unsigned long timeout;
 	int ret;
 
 	/* Select the right mux port and perform the transfer. */
 
+	hold_msg = i2c_check_hold_msg(flags,
+				      size == I2C_SMBUS_WORD_DATA ? 2 : 0,
+				      &data->word);
+	if (hold_msg == I2C_HOLD_MSG_SET) {
+		timeout = msecs_to_jiffies(data->word);
+		i2c_mux_hold(muxc, priv->chan_id, timeout);
+	} else if (hold_msg == I2C_HOLD_MSG_NONE) {
+		mutex_lock(&muxc->hold_lock);
+	}
+
 	ret = muxc->select(muxc, priv->chan_id);
 	if (ret >= 0)
 		ret = i2c_smbus_xfer(parent, addr, flags,
-				     read_write, command, size, data);
-	if (muxc->deselect)
-		muxc->deselect(muxc, priv->chan_id);
+				 read_write, command, size, data);
+	if (hold_msg != I2C_HOLD_MSG_SET) {
+		if (muxc->deselect)
+			muxc->deselect(muxc, priv->chan_id);
+		if (hold_msg == I2C_HOLD_MSG_RESET)
+			i2c_mux_unhold(muxc);
+		else
+			mutex_unlock(&muxc->hold_lock);
+	}
 
 	return ret;
 }
@@ -267,10 +365,10 @@ struct i2c_adapter *i2c_root_adapter(struct device *dev)
 EXPORT_SYMBOL_GPL(i2c_root_adapter);
 
 struct i2c_mux_core *i2c_mux_alloc(struct i2c_adapter *parent,
-				   struct device *dev, int max_adapters,
-				   int sizeof_priv, u32 flags,
-				   int (*select)(struct i2c_mux_core *, u32),
-				   int (*deselect)(struct i2c_mux_core *, u32))
+			   struct device *dev, int max_adapters,
+			   int sizeof_priv, u32 flags,
+			   int (*select)(struct i2c_mux_core *, u32),
+			   int (*deselect)(struct i2c_mux_core *, u32))
 {
 	struct i2c_mux_core *muxc;
 	size_t mux_size;
@@ -291,6 +389,9 @@ struct i2c_mux_core *i2c_mux_alloc(struct i2c_adapter *parent,
 	muxc->deselect = deselect;
 	muxc->max_adapters = max_adapters;
 
+	mutex_init(&muxc->hold_lock);
+	INIT_DELAYED_WORK(&muxc->unhold_work, i2c_mux_unhold_work);
+
 	return muxc;
 }
 EXPORT_SYMBOL_GPL(i2c_mux_alloc);
@@ -412,7 +513,11 @@ int i2c_mux_add_adapter(struct i2c_mux_core *muxc,
 			}
 		}
 
-		priv->adap.dev.of_node = child;
+		if (child)
+			priv->adap.dev.of_node = child;
+		else
+			priv->adap.dev.of_node = mux_node;
+
 		of_node_put(mux_node);
 	}
 
@@ -467,6 +572,8 @@ void i2c_mux_del_adapters(struct i2c_mux_core *muxc)
 {
 	char symlink_name[20];
 
+	i2c_mux_unhold(muxc);
+	
 	while (muxc->num_adapters) {
 		struct i2c_adapter *adap = muxc->adapter[--muxc->num_adapters];
 		struct i2c_mux_priv *priv = adap->algo_data;
diff --git a/drivers/i2c/muxes/i2c-arb-gpio-challenge.c b/drivers/i2c/muxes/i2c-arb-gpio-challenge.c
index d6ef91b888c6..6c2cb1c5fecc 100644
--- a/drivers/i2c/muxes/i2c-arb-gpio-challenge.c
+++ b/drivers/i2c/muxes/i2c-arb-gpio-challenge.c
@@ -13,7 +13,10 @@
 #include <linux/module.h>
 #include <linux/platform_device.h>
 #include <linux/slab.h>
-
+#if CONFIG_DRAGON_CHASSIS
+#include <linux/device.h>
+#include <linux/timer.h>
+#endif //CONFIG_DRAGON_CHASSIS
 
 /**
  * struct i2c_arbitrator_data - Driver data for I2C arbitrator
@@ -28,9 +31,19 @@
 struct i2c_arbitrator_data {
 	struct gpio_desc *our_gpio;
 	struct gpio_desc *their_gpio;
+#if CONFIG_DRAGON_CHASSIS
+	struct gpio_desc *cpld_refresh;
+#endif
 	unsigned int slew_delay_us;
 	unsigned int wait_retry_us;
 	unsigned int wait_free_us;
+#if CONFIG_DRAGON_CHASSIS
+	struct timer_list timer;
+	struct i2c_mux_core *muxc;
+	struct i2c_adapter *cpld_access;
+	u8 cpld_address;
+	bool bus_locked;
+#endif //CONFIG_DRAGON_CHASSIS
 };
 
 
@@ -43,7 +56,14 @@ static int i2c_arbitrator_select(struct i2c_mux_core *muxc, u32 chan)
 {
 	const struct i2c_arbitrator_data *arb = i2c_mux_priv(muxc);
 	unsigned long stop_retry, stop_time;
-
+#if CONFIG_DRAGON_CHASSIS
+	if (!IS_ERR(arb->cpld_refresh)) {
+		if (gpiod_get_value(arb->cpld_refresh))
+			return -EBUSY;
+	}
+	if (arb->bus_locked)
+		return 0;
+#endif //CONFIG_DRAGON_CHASSIS
 	/* Start a round of trying to claim the bus */
 	stop_time = jiffies + usecs_to_jiffies(arb->wait_free_us) + 1;
 	do {
@@ -85,7 +105,10 @@ static int i2c_arbitrator_select(struct i2c_mux_core *muxc, u32 chan)
 static int i2c_arbitrator_deselect(struct i2c_mux_core *muxc, u32 chan)
 {
 	const struct i2c_arbitrator_data *arb = i2c_mux_priv(muxc);
-
+#if CONFIG_DRAGON_CHASSIS
+	if (arb->bus_locked) //if bus is locked do not deselect it
+		return 0;
+#endif //CONFIG_DRAGON_CHASSIS
 	/* Release the bus and wait for the other master to notice */
 	gpiod_set_value(arb->our_gpio, 0);
 	udelay(arb->slew_delay_us);
@@ -93,6 +116,105 @@ static int i2c_arbitrator_deselect(struct i2c_mux_core *muxc, u32 chan)
 	return 0;
 }
 
+#if CONFIG_DRAGON_CHASSIS
+static ssize_t cpld_refresh_show(struct device *dev,
+				 struct device_attribute *attr,
+				 char *buf)
+{
+	struct i2c_client *client = to_i2c_client(dev);
+	struct i2c_mux_core *muxc = i2c_get_clientdata(client);
+	const struct i2c_arbitrator_data *arb = i2c_mux_priv(muxc);
+	int gpio_val = -1;
+	if (!IS_ERR(arb->cpld_refresh))
+		gpio_val = gpiod_get_value(arb->cpld_refresh);
+	return sprintf(buf, "%d\n", gpio_val);
+}
+static DEVICE_ATTR_RO(cpld_refresh);
+
+static ssize_t arb_state_show(struct device *dev,
+			    struct device_attribute *attr,
+			    char *buf)
+{
+	struct i2c_client *client = to_i2c_client(dev);
+	struct i2c_mux_core *muxc = i2c_get_clientdata(client);
+	const struct i2c_arbitrator_data *arb = i2c_mux_priv(muxc);
+	int gpio_val = gpiod_get_value(arb->their_gpio);
+	if (!IS_ERR(arb->cpld_refresh)) {
+		if (gpiod_get_value(arb->cpld_refresh))
+			return -EBUSY;
+	}
+	return sprintf(buf, "%d\n", gpio_val);
+}
+
+static ssize_t arb_state_store(struct device *dev,
+			     struct device_attribute *attr,
+			     const char *buf, size_t count)
+{
+	struct i2c_client *client = to_i2c_client(dev);
+	struct i2c_mux_core *muxc = i2c_get_clientdata(client);
+	struct i2c_arbitrator_data *arb = i2c_mux_priv(muxc);
+	int val;
+	int ret;
+	ret = kstrtoint(buf, 0, &val);
+	if (ret < 0)
+		return ret;
+	if (val != 0 && val != 1)
+		return -EINVAL;
+	if (!val) {
+		arb->bus_locked = false;
+		del_timer_sync(&arb->timer);
+		i2c_arbitrator_deselect(muxc, 0);
+	} else {
+		ret = i2c_arbitrator_select(muxc, 0);
+		if (ret < 0)
+			return ret;
+		arb->timer.expires = jiffies + HZ;
+		add_timer(&arb->timer);
+		arb->bus_locked = true;
+	}
+	return count;
+}
+static DEVICE_ATTR_RW(arb_state);
+
+static void pet_cpld_register(struct timer_list *t)
+{
+	struct i2c_arbitrator_data *arb = from_timer(arb, t, timer);
+	struct i2c_mux_core *muxc = arb->muxc;
+	int gpio_val;
+	u8 out_buf0[2];
+	u8 out_buf1[2];
+	struct i2c_msg msgs[] = {
+		{
+			.addr = arb->cpld_address,
+			.flags = 0,
+			.len = 2,
+			.buf = out_buf0,
+		},
+		{
+			.addr = arb->cpld_address,
+			.flags = 0,
+			.len = 2,
+			.buf = out_buf1,
+		}
+	};
+	out_buf0[0] = 16;
+	out_buf0[1] = 0x5;
+	out_buf1[0] = 16;
+	out_buf1[1] = 0xA;
+	gpio_val = gpiod_get_value(arb->their_gpio);
+	if (!gpio_val)
+	{
+		if (i2c_transfer(arb->cpld_access, msgs, 2) != 2)
+			dev_err(muxc->dev, "Could not access cpld to pet the watchdog\n");
+		mod_timer(&arb->timer, jiffies + HZ);
+	}
+	else
+	{
+		dev_err(muxc->dev, "Unexpected gpio value in timer.\n");
+	}
+}
+#endif //CONFIG_DRAGON_CHASSIS
+
 static int i2c_arbitrator_probe(struct platform_device *pdev)
 {
 	struct device *dev = &pdev->dev;
@@ -118,7 +240,9 @@ static int i2c_arbitrator_probe(struct platform_device *pdev)
 	if (!muxc)
 		return -ENOMEM;
 	arb = i2c_mux_priv(muxc);
-
+#if CONFIG_DRAGON_CHASSIS
+	arb->muxc = muxc;
+#endif //CONFIG_DRAGON_CHASSIS
 	platform_set_drvdata(pdev, muxc);
 
 	/* Request GPIOs, our GPIO as unclaimed to begin with */
diff --git a/include/linux/device/class.h b/include/linux/device/class.h
index 65880e60c720..f04f4536bc8a 100644
--- a/include/linux/device/class.h
+++ b/include/linux/device/class.h
@@ -226,4 +226,27 @@ void class_interface_unregister(struct class_interface *);
 struct class * __must_check class_create(const char *name);
 void class_destroy(const struct class *cls);
 
+/* This is a #define to keep the compiler from merging different
+ * instances of the __key variable */
+
+/**
+ * class_create - create a struct class structure
+ * @owner: pointer to the module that is to "own" this struct class
+ * @name: pointer to a string for the name of this class.
+ *
+ * This is used to create a struct class pointer that can then be used
+ * in calls to device_create().
+ *
+ * Returns &struct class pointer on success, or ERR_PTR() on error.
+ *
+ * Note, the pointer created here is to be destroyed when finished by
+ * making a call to class_destroy().
+ */
+/*#define class_create(owner, name)               \
+({                                              \
+        static struct lock_class_key __key;     \
+        __class_create(owner, name, &__key);    \
+})
+*/
+
 #endif	/* _DEVICE_CLASS_H_ */
diff --git a/include/linux/i2c-aspeed.h b/include/linux/i2c-aspeed.h
new file mode 100644
index 000000000000..57d47d2ea82c
--- /dev/null
+++ b/include/linux/i2c-aspeed.h
@@ -0,0 +1,216 @@
+#ifndef I2C_ASPEED_HDR
+#define I2C_ASPEED_HDR
+
+#include <linux/i2c.h>
+#include <linux/io.h>
+
+/* I2C Global Registers */
+/* 0x00 : I2CG Interrupt Status Register  */
+/* 0x08 : I2CG Interrupt Target Assignment  */
+/* 0x0c : I2CG Global Control Register (AST2500)  */
+#define ASPEED_I2CG_GLOBAL_CTRL_REG			0x0c
+#define  ASPEED_I2CG_SRAM_BUFFER_EN			BIT(0)
+
+/* I2C Bus Registers */
+#define ASPEED_I2C_FUN_CTRL_REG				0x00
+#define ASPEED_I2C_AC_TIMING_REG1			0x04
+#define ASPEED_I2C_AC_TIMING_REG2			0x08
+#define ASPEED_I2C_INTR_CTRL_REG			0x0c
+#define ASPEED_I2C_INTR_STS_REG				0x10
+#define ASPEED_I2C_CMD_REG				0x14
+#define ASPEED_I2C_DEV_ADDR_REG				0x18
+#define ASPEED_I2C_BUF_CTRL_REG				0x1c
+#define ASPEED_I2C_BYTE_BUF_REG				0x20
+#define ASPEED_I2C_DMA_ADDR_REG				0x24
+#define ASPEED_I2C_DMA_LEN_REG				0x28
+
+/* Device Register Definition */
+/* 0x00 : I2CD Function Control Register  */
+#define ASPEED_I2CD_BUFFER_PAGE_SEL_MASK		GENMASK(22, 20)
+#define ASPEED_I2CD_BUS_AUTO_RECOVERY_EN		BIT(17)
+#define ASPEED_I2CD_MULTI_MASTER_DIS			BIT(15)
+#define ASPEED_I2CD_SDA_DRIVE_1T_EN			BIT(8)
+#define ASPEED_I2CD_M_SDA_DRIVE_1T_EN			BIT(7)
+#define ASPEED_I2CD_M_HIGH_SPEED_EN			BIT(6)
+#define ASPEED_I2CD_GCALL_EN				BIT(2)
+#define ASPEED_I2CD_SLAVE_EN				BIT(1)
+#define ASPEED_I2CD_MASTER_EN				BIT(0)
+
+/* 0x04 : I2CD Clock and AC Timing Control Register #1 */
+#define ASPEED_I2CD_TIME_TBUF_MASK			GENMASK(31, 28)
+#define ASPEED_I2CD_TIME_THDSTA_MASK			GENMASK(27, 24)
+#define ASPEED_I2CD_TIME_TACST_MASK			GENMASK(23, 20)
+#define ASPEED_I2CD_TIME_SCL_HIGH_SHIFT			16
+#define ASPEED_I2CD_TIME_SCL_HIGH_MASK			GENMASK(19, 16)
+#define ASPEED_I2CD_TIME_SCL_LOW_SHIFT			12
+#define ASPEED_I2CD_TIME_SCL_LOW_MASK			GENMASK(15, 12)
+#define ASPEED_I2CD_TIME_TIMEOUT_BASE_DIVISOR_SHIFT	8
+#define ASPEED_I2CD_TIME_TIMEOUT_BASE_DIVISOR_MASK	GENMASK(9, 8)
+#define ASPEED_I2CD_TIME_BASE_DIVISOR_MASK		GENMASK(3, 0)
+#define ASPEED_I2CD_TIME_SCL_REG_MAX			GENMASK(3, 0)
+
+/* 0x08 : I2CD Clock and AC Timing Control Register #2 */
+#define ASPEED_I2CD_TIMEOUT_CYCLES_SHIFT		0
+#define ASPEED_I2CD_TIMEOUT_CYCLES_MASK			GENMASK(4, 0)
+
+/* 0x0c : I2CD Interrupt Control Register &
+ * 0x10 : I2CD Interrupt Status Register
+ *
+ * These share bit definitions, so use the same values for the enable &
+ * status bits.
+ */
+#define ASPEED_I2CD_INTR_RECV_MASK			0xf000ffff
+#if defined(CONFIG_MACH_ASPEED_G6)
+#define ASPEED_I2CD_INTR_SLAVE_ADDR_RECEIVED_PENDING	BIT(29)
+#else
+#define ASPEED_I2CD_INTR_SLAVE_ADDR_RECEIVED_PENDING	BIT(30)
+#endif
+#define ASPEED_I2CD_INTR_SLAVE_INACTIVE_TIMEOUT		BIT(15)
+#define ASPEED_I2CD_INTR_SDA_DL_TIMEOUT			BIT(14)
+#define ASPEED_I2CD_INTR_BUS_RECOVER_DONE		BIT(13)
+#define ASPEED_I2CD_INTR_GCALL_ADDR			BIT(8)
+#define ASPEED_I2CD_INTR_SLAVE_MATCH			BIT(7)
+#define ASPEED_I2CD_INTR_SCL_TIMEOUT			BIT(6)
+#define ASPEED_I2CD_INTR_ABNORMAL			BIT(5)
+#define ASPEED_I2CD_INTR_NORMAL_STOP			BIT(4)
+#define ASPEED_I2CD_INTR_ARBIT_LOSS			BIT(3)
+#define ASPEED_I2CD_INTR_RX_DONE			BIT(2)
+#define ASPEED_I2CD_INTR_TX_NAK				BIT(1)
+#define ASPEED_I2CD_INTR_TX_ACK				BIT(0)
+#define ASPEED_I2CD_INTR_MASTER_ERRORS					       \
+		(ASPEED_I2CD_INTR_SDA_DL_TIMEOUT |			       \
+		 ASPEED_I2CD_INTR_SCL_TIMEOUT |				       \
+		 ASPEED_I2CD_INTR_ABNORMAL |				       \
+		 ASPEED_I2CD_INTR_ARBIT_LOSS)
+#define ASPEED_I2CD_INTR_SLAVE_ERRORS					       \
+		ASPEED_I2CD_INTR_SLAVE_INACTIVE_TIMEOUT
+#define ASPEED_I2CD_INTR_ALL						       \
+		(ASPEED_I2CD_INTR_SLAVE_INACTIVE_TIMEOUT |		       \
+		 ASPEED_I2CD_INTR_SDA_DL_TIMEOUT |			       \
+		 ASPEED_I2CD_INTR_BUS_RECOVER_DONE |			       \
+		 ASPEED_I2CD_INTR_SCL_TIMEOUT |				       \
+		 ASPEED_I2CD_INTR_ABNORMAL |				       \
+		 ASPEED_I2CD_INTR_NORMAL_STOP |				       \
+		 ASPEED_I2CD_INTR_ARBIT_LOSS |				       \
+		 ASPEED_I2CD_INTR_RX_DONE |				       \
+		 ASPEED_I2CD_INTR_TX_NAK |				       \
+		 ASPEED_I2CD_INTR_TX_ACK)
+#define ASPEED_I2CD_INTR_STATUS_MASK					       \
+		(ASPEED_I2CD_INTR_SLAVE_ADDR_RECEIVED_PENDING |		       \
+		 ASPEED_I2CD_INTR_GCALL_ADDR |				       \
+		 ASPEED_I2CD_INTR_SLAVE_MATCH |				       \
+		 ASPEED_I2CD_INTR_ALL)
+
+/* 0x14 : I2CD Command/Status Register   */
+#define ASPEED_I2CD_SCL_LINE_STS			BIT(18)
+#define ASPEED_I2CD_SDA_LINE_STS			BIT(17)
+#define ASPEED_I2CD_BUS_BUSY_STS			BIT(16)
+#define ASPEED_I2CD_BUS_RECOVER_CMD			BIT(11)
+
+/* Command Bit */
+#define ASPEED_I2CD_RX_DMA_ENABLE			BIT(9)
+#define ASPEED_I2CD_TX_DMA_ENABLE			BIT(8)
+#define ASPEED_I2CD_RX_BUFF_ENABLE			BIT(7)
+#define ASPEED_I2CD_TX_BUFF_ENABLE			BIT(6)
+#define ASPEED_I2CD_M_STOP_CMD				BIT(5)
+#define ASPEED_I2CD_M_S_RX_CMD_LAST			BIT(4)
+#define ASPEED_I2CD_M_RX_CMD				BIT(3)
+#define ASPEED_I2CD_S_TX_CMD				BIT(2)
+#define ASPEED_I2CD_M_TX_CMD				BIT(1)
+#define ASPEED_I2CD_M_START_CMD				BIT(0)
+#define ASPEED_I2CD_MASTER_CMDS_MASK					       \
+		(ASPEED_I2CD_M_STOP_CMD |				       \
+		 ASPEED_I2CD_M_S_RX_CMD_LAST |				       \
+		 ASPEED_I2CD_M_RX_CMD |					       \
+		 ASPEED_I2CD_M_TX_CMD |					       \
+		 ASPEED_I2CD_M_START_CMD)
+
+/* 0x18 : I2CD Slave Device Address Register   */
+#define ASPEED_I2CD_DEV_ADDR_MASK			GENMASK(6, 0)
+
+/* 0x1c : I2CD Buffer Control Register */
+/* Use 8-bits or 6-bits wide bit fileds to support both AST2400 and AST2500 */
+#define ASPEED_I2CD_BUF_RX_COUNT_MASK			GENMASK(31, 24)
+#define ASPEED_I2CD_BUF_RX_SIZE_MASK			GENMASK(23, 16)
+#define ASPEED_I2CD_BUF_TX_COUNT_MASK			GENMASK(15, 8)
+#define ASPEED_I2CD_BUF_OFFSET_MASK			GENMASK(5, 0)
+
+/* 0x24 : I2CD DMA Mode Buffer Address Register */
+#define ASPEED_I2CD_DMA_ADDR_MASK			GENMASK(31, 2)
+#define ASPEED_I2CD_DMA_ALIGN				4
+
+/* 0x28 : I2CD DMA Transfer Length Register */
+#define ASPEED_I2CD_DMA_LEN_SHIFT			0
+#define ASPEED_I2CD_DMA_LEN_MASK			GENMASK(11, 0)
+
+enum aspeed_i2c_master_state {
+	ASPEED_I2C_MASTER_INACTIVE,
+	ASPEED_I2C_MASTER_PENDING,
+	ASPEED_I2C_MASTER_START,
+	ASPEED_I2C_MASTER_TX_FIRST,
+	ASPEED_I2C_MASTER_TX,
+	ASPEED_I2C_MASTER_RX_FIRST,
+	ASPEED_I2C_MASTER_RX,
+	ASPEED_I2C_MASTER_STOP,
+};
+
+enum aspeed_i2c_slave_state {
+	ASPEED_I2C_SLAVE_INACTIVE,
+	ASPEED_I2C_SLAVE_START,
+	ASPEED_I2C_SLAVE_READ_REQUESTED,
+	ASPEED_I2C_SLAVE_READ_PROCESSED,
+	ASPEED_I2C_SLAVE_WRITE_REQUESTED,
+	ASPEED_I2C_SLAVE_WRITE_RECEIVED,
+	ASPEED_I2C_SLAVE_GCALL_START,
+	ASPEED_I2C_SLAVE_GCALL_REQUESTED,
+	ASPEED_I2C_SLAVE_STOP,
+};
+
+struct aspeed_i2c_bus {
+	struct i2c_adapter		adap;
+	struct device			*dev;
+	void __iomem			*base;
+	struct reset_control		*rst;
+	/* Synchronizes I/O mem access to base. */
+	spinlock_t			lock;
+	struct completion		cmd_complete;
+	u32				(*get_clk_reg_val)(struct device *dev,
+							   u32 divisor);
+	unsigned long			parent_clk_frequency;
+	u32				bus_frequency;
+	u32				hw_timeout_ms;
+	/* Transaction state. */
+	enum aspeed_i2c_master_state	master_state;
+	struct i2c_msg			*msgs;
+	size_t				buf_index;
+	size_t				msgs_index;
+	size_t				msgs_count;
+	bool				send_stop;
+	int				cmd_err;
+	/* Protected only by i2c_lock_bus */
+	int				master_xfer_result;
+	/* Multi-master */
+	bool				multi_master;
+	/* Buffer mode */
+	void __iomem			*buf_base;
+	u8				buf_offset;
+	u8				buf_page;
+	/* DMA mode */
+	struct dma_pool			*dma_pool;
+	dma_addr_t			dma_handle;
+	u8				*dma_buf;
+	size_t				dma_len;
+	/* Buffer/DMA mode */
+	size_t				buf_size;
+#if IS_ENABLED(CONFIG_I2C_SLAVE)
+	struct i2c_client		*slave;
+	enum aspeed_i2c_slave_state	slave_state;
+	/* General call */
+	bool				general_call;
+#endif /* CONFIG_I2C_SLAVE */
+};
+
+/* Per ipmi_ssif driver, maximum retry count for a message is 250 */
+#define SSIF_RETRIES            250
+
+#endif
diff --git a/include/linux/i2c-mux.h b/include/linux/i2c-mux.h
index 1784ac7afb11..476ffcaf7247 100644
--- a/include/linux/i2c-mux.h
+++ b/include/linux/i2c-mux.h
@@ -14,6 +14,7 @@
 #ifdef __KERNEL__
 
 #include <linux/bitops.h>
+#include <linux/workqueue.h>
 
 struct i2c_mux_core {
 	struct i2c_adapter *parent;
@@ -27,6 +28,10 @@ struct i2c_mux_core {
 	int (*select)(struct i2c_mux_core *, u32 chan_id);
 	int (*deselect)(struct i2c_mux_core *, u32 chan_id);
 
+	struct mutex hold_lock; /* mutex for channel holding */
+	u32 holder_chan_id;
+	struct delayed_work unhold_work;
+
 	int num_adapters;
 	int max_adapters;
 	struct i2c_adapter *adapter[];
diff --git a/include/linux/i2c.h b/include/linux/i2c.h
index 45431a3a3205..c971a7233cd2 100644
--- a/include/linux/i2c.h
+++ b/include/linux/i2c.h
@@ -784,6 +784,13 @@ struct i2c_adapter {
 	struct irq_domain *host_notify_domain;
 	struct regulator *bus_regulator;
 
+	/*
+	 * These will be used by root adpaters only. For muxes, each mux core
+	 * has these individually.
+	 */
+	struct mutex hold_lock; /* mutex for bus holding */
+	struct delayed_work unhold_work;
+
 	struct dentry *debugfs;
 
 	/* 7bit address space */
@@ -1118,7 +1125,7 @@ static inline int of_i2c_get_board_info(struct device *dev,
 struct acpi_resource;
 struct acpi_resource_i2c_serialbus;
 
-#if IS_REACHABLE(CONFIG_ACPI) && IS_REACHABLE(CONFIG_I2C)
+#if IS_REACHABLE(CONFIG_ACPI)
 bool i2c_acpi_get_i2c_resource(struct acpi_resource *ares,
 			       struct acpi_resource_i2c_serialbus **i2c);
 int i2c_acpi_client_count(struct acpi_device *adev);
@@ -1165,4 +1172,23 @@ static inline struct i2c_client *i2c_acpi_new_device(struct device *dev,
 	return i2c_acpi_new_device_by_fwnode(dev_fwnode(dev), index, info);
 }
 
+enum i2c_hold_msg_type {
+        I2C_HOLD_MSG_NONE,
+        I2C_HOLD_MSG_SET,
+        I2C_HOLD_MSG_RESET
+};
+
+static inline enum i2c_hold_msg_type i2c_check_hold_msg(u16 flags, u16 len, u16 *buf)
+{
+	if (flags & I2C_M_HOLD && len == sizeof(u16)) {
+		if (*buf)
+			return I2C_HOLD_MSG_SET;
+
+		return I2C_HOLD_MSG_RESET;
+	}
+
+	return I2C_HOLD_MSG_NONE;
+}
+
+
 #endif /* _LINUX_I2C_H */
diff --git a/include/uapi/linux/i2c.h b/include/uapi/linux/i2c.h
index a2db2a56c8b0..1766411f7947 100644
--- a/include/uapi/linux/i2c.h
+++ b/include/uapi/linux/i2c.h
@@ -76,6 +76,7 @@ struct i2c_msg {
 	__u16 flags;
 #define I2C_M_RD		0x0001	/* guaranteed to be 0x0001! */
 #define I2C_M_TEN		0x0010	/* use only if I2C_FUNC_10BIT_ADDR */
+#define I2C_M_HOLD		0x0100	/* for holding a mux path */
 #define I2C_M_DMA_SAFE		0x0200	/* use only in kernel space */
 #define I2C_M_RECV_LEN		0x0400	/* use only if I2C_FUNC_SMBUS_READ_BLOCK_DATA */
 #define I2C_M_NO_RD_ACK		0x0800	/* use only if I2C_FUNC_PROTOCOL_MANGLING */
-- 
2.34.1

